### ISSUE TITLE
NVDA unable to fetch add-ons in add-on store when a corporate HTTPS certificate is present

### ISSUE THREAD

**Issue by kyleman** [nvaccess/nvda#15905]
### Steps to reproduce:
Fetch any add-on data in the add-ons store.

### Actual behavior:
NVDA fails with an error popup, "Add-on data update failure unable to fetch latest add-on data for compatible add-ons."

### Expected behavior:
There is no error and add-on data is able to be fetched normally.

### NVDA logs, crash dumps and other attachments:
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._getIA2RelationFirstTarget (14:02:18.283) - MainThread (14060):
Unable to use _getIA2TargetsForRelationsOfType, fallback to _IA2Relations.
DEBUGWARNING - _addonStore.dataManager._DataManager._getCacheHash (14:02:18.611) - getAddonData (10688):
Unable to get cache hash: HTTPSConnectionPool(host='nvaccess.org', port=443): Max retries exceeded with url: /addonStore/cacheHash.json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)')))
DEBUGWARNING - _addonStore.dataManager._DataManager._getLatestAddonsDataForVersion (14:02:18.821) - getAddonData (10688):
Unable to fetch addon data: HTTPSConnectionPool(host='nvaccess.org', port=443): Max retries exceeded with url: /addonStore/en/all/2023.3.0.json (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1091)')))

#### NVDA installed/portable/running from source:
portable NVDA

#### NVDA version:
2023.3rc2 

#### Windows version:
Windows 10 22H2
Version; 10.0.19045Build 19045

#### Name and version of other software in use when reproducing the issue:
N/A

#### Other information about your system:
N/A

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yess all versions since the add-ons store has been added.

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by Adriani90** [comment]
Is this also happening when you try to fetch updates? i.e. when you update NVDA to an alpha version for example?

This issue might be related to #5871 which is still reproducible.

There is also a python issue that might be related to this:
https://github.com/python/cpython/issues/65115

**Comment by kyleman** [comment]
No I am able to download nvda updates normally without triggering any errors.

**Comment by kyleman** [comment]
Just commenting hear that this also still doesn't work in 2024.1 beta one or beta two.

**Comment by Neurrone** [comment]
I'm also encountering this. It seems to not affect updating of NVDA itself though, just add-on store updates.

**Comment by hwf1324** [comment]
Hi, @seanbudd 

I don't know how to make requests trust the system Trusted Root Certification Authorities list, can we add a configuration option to disable SSL validation?

I know it's not secure, but it's the only thing I can do.

Thanks.

**Comment by seanbudd** [comment]
I think you may need to install NVDA for us to be able to update root certificates - can you try with an installed copy of NVDA first and then try running the portable?

**Comment by hwf1324** [comment]
> I think you may need to install NVDA for us to be able to update root certificates - can you try with an installed copy of NVDA first and then try running the portable?

I'm a little unsure of the meaning of this passage.

Meaning that when NVDA is installed, NVDA gets the certificate from the system's Trusted Root Certification Authorities list?

As I understand it, Requests uses certificates from the package certifi.

And certifi does not have the ability for users to manually add certificates.

---

Maybe the attempts I've made recently have caused some issues with getting the add-on's cached data to prompt me that I'm not connected to the internet.

This hasn't happened before, and it was normal again some time after I restored the environment variables.

(Caused by SSLError(SSLError(136, '[X509: NO_CERTIFICATE_OR_CRL_FOUND] no certificate or crl found (_ssl.c:4166)')))

The long term problem encountered is the following error when downloading GItHub resources. I am still encountering this error in my newly created copy of the portable version.

(Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')))

**Comment by Adriani90** [comment]
@kyleman, @Neurrone, @hwf1324 are you having this issue also in NVDA 2025.1 Beta on Windows 11 24 H2 last stable update?
Which browser is your default browser for http and https link types (opening links in background)? Can you change the default browser in Windows for the coresponding link types? Does that have any influence on the issue?

**Comment by hwf1324** [comment]
> [@kyleman](https://github.com/kyleman), [@Neurrone](https://github.com/Neurrone), [@hwf1324](https://github.com/hwf1324) are you having this issue also in NVDA 2025.1 Beta on Windows 11 24 H2 last stable update? Which browser is your default browser for http and https link types (opening links in background)? Can you change the default browser in Windows for the coresponding link types? Does that have any influence on the issue?

Although I have not tested it, I believe this issue still exists.

However, I got around it by adding the required trusted certificates to the certifi copy and using this copy via an environment variable.

Below is an article I wrote in Chinese that implicitly documents the method I used at the end.

https://www.nvdacn.com/index.php/archives/1393/

**Comment by seanbudd** [comment]
Fixed by #18354 

@kyleman - can you please test the latest [alpha](https://download.nvaccess.org/snapshots/alpha/) or the upcoming 2025.2beta1 to confirm? The beta should be released next Tuesday

**Comment by hwf1324** [comment]
This issue has not been fixed in version alpha-37061,9e668b18 (2025.3.0.37061).

```log
DEBUGWARNING - addonStore.network.AddonFileDownloader._download (09:37:05.809) - AddonDownloader_0 (37516):
Unable to download addon file: HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /mltony/nvda-console-toolkit/releases/download/v1.4.5/consoleToolkit-1.4.5.nvda-addon (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')))
DEBUG - addonStore.network.AddonFileDownloader._done (09:37:05.809) - AddonDownloader_0 (37516):
Done called for consoleToolkit
DEBUG - gui.addonStoreGui.viewModels.addonList.AddonListItemVM.status (09:37:05.809) - AddonDownloader_0 (37516):
addon status change: consoleToolkit-Channel.STABLE: status: AvailableAddonStatus.DOWNLOAD_FAILED
DEBUGWARNING - gui.addonStoreGui.viewModels.store.AddonStoreVM._downloadComplete (09:37:05.810) - AddonDownloader_0 (37516):
Error during download of consoleToolkit-Channel.STABLE
Traceback (most recent call last):
  File "urllib3\connectionpool.pyc", line 464, in _make_request
  File "urllib3\connectionpool.pyc", line 1093, in _validate_conn
  File "urllib3\connection.pyc", line 741, in connect
  File "urllib3\connection.pyc", line 920, in _ssl_wrap_socket_and_match_hostname
  File "urllib3\util\ssl_.pyc", line 480, in ssl_wrap_socket
  File "urllib3\util\ssl_.pyc", line 524, in _ssl_wrap_socket_impl
  File "ssl.pyc", line 517, in wrap_socket
  File "ssl.pyc", line 1104, in _create
  File "ssl.pyc", line 1382, in do_handshake
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "urllib3\connectionpool.pyc", line 787, in urlopen
  File "urllib3\connectionpool.pyc", line 488, in _make_request
urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "requests\adapters.pyc", line 667, in send
  File "urllib3\connectionpool.pyc", line 841, in urlopen
  File "urllib3\util\retry.pyc", line 519, in increment
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /mltony/nvda-console-toolkit/releases/download/v1.4.5/consoleToolkit-1.4.5.nvda-addon (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "addonStore\network.pyc", line 258, in _download
  File "addonStore\network.pyc", line 219, in _downloadAddonToPath
  File "requests\api.pyc", line 73, in get
  File "requests\api.pyc", line 59, in request
  File "requests\sessions.pyc", line 589, in request
  File "requests\sessions.pyc", line 703, in send
  File "requests\adapters.pyc", line 698, in send
requests.exceptions.SSLError: HTTPSConnectionPool(host='github.com', port=443): Max retries exceeded with url: /mltony/nvda-console-toolkit/releases/download/v1.4.5/consoleToolkit-1.4.5.nvda-addon (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1006)')))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "concurrent\futures\thread.pyc", line 58, in run
  File "addonStore\network.pyc", line 262, in _download
gui.message.DisplayableError: ('插件 Console Toolkit 无法下载', '插件下载失败')
```

**Comment by seanbudd** [comment]
@hwf1324 - this issue about fetching add-ons in the add-on store, not downloading specific add-ons.
I'll look into fixing this for downloading add-ons from GitHub, but I don't feel comfortable adding code to trust certificates from other websites.

**Comment by hwf1324** [comment]
Sorry, I've always confused these two issues.

**Comment by seanbudd** [comment]
@hwf1324 - could you test the PR build for https://github.com/nvaccess/nvda/pull/18402 when it completes? This should give you an option to trust the GitHub cert and download the add-on

**Comment by kyleman** [comment]
So what did this exactly break now that it has been removed and delayed until next year? It worked fine under beta four, but no longer on beta five.

**Comment by seanbudd** [comment]
This broke add-ons, it inadvertently broke the add-on API. We will fix this in 2026.1

**Comment by seanbudd** [comment]
If this is to be done, we want it to be done in an API safe way.
I think we can consider removing this from the 2026.1 milestone to reduce the scope as per https://github.com/nvaccess/nvda/pull/19106#issue-3516807604
